process/txsSender/txsSender.go

```diff
@@ -0,0 +1,237 @@
+// SendBulkTransactions sends the provided transactions as a bulk, optimizing transfer between nodes
+func (ts *txsSender) SendBulkTransactions(txs []*transaction.Transaction) (uint64, error) {
+	if len(txs) == 0 {
+		return 0, process.ErrNoTxToProcess
+	}
+
+	ts.addTransactionsToSendPipe(txs)
+
+	return uint64(len(txs)), nil
+}
+
+func (ts *txsSender) addTransactionsToSendPipe(txs []*transaction.Transaction) {
+	for _, tx := range txs {
+		ts.txAccumulator.AddData(tx)
+	}
+}
+
+func (ts *txsSender) sendFromTxAccumulator(ctx context.Context) {
+	outputChannel := ts.txAccumulator.OutputChannel()
+
+	for {
+		select {
+		case objs := <-outputChannel:
+			{
+				if len(objs) == 0 {
+					break
+				}
+
+				txs := make([]*transaction.Transaction, 0, len(objs))
+				for _, obj := range objs {
+					tx, ok := obj.(*transaction.Transaction)
+					if !ok {
+						continue
+					}
+
+					txs = append(txs, tx)
+				}
+
+				atomic.AddUint32(&ts.txSentCounter, uint32(len(txs)))
+
+				ts.sendBulkTransactions(txs)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (ts *txsSender) sendBulkTransactions(txs []*transaction.Transaction) {
+	transactionsByShards := make(map[uint32][][]byte)
+	log.Trace("txsSender.sendBulkTransactions sending txs",
+		"num", len(txs),
+	)
+
+	for _, tx := range txs {
+		marshalledTx, err := ts.marshaller.Marshal(tx)
+		if err != nil {
+			log.Warn("txsSender.sendBulkTransactions",
+				"marshaller error", err,
+			)
+			continue
+		}
+
+		senderShardId := ts.shardCoordinator.ComputeId(tx.SndAddr)
+		transactionsByShards[senderShardId] = append(transactionsByShards[senderShardId], marshalledTx)
+	}
+
+	for shardId, txsForShard := range transactionsByShards {
+		err := ts.sendBulkTransactionsFromShard(txsForShard, shardId)
+		log.LogIfError(err)
+	}
+}
+
+func (ts *txsSender) sendBulkTransactionsFromShard(transactions [][]byte, senderShardId uint32) error {
+	// the topic identifier is made of the current shard id and sender's shard id
+	identifier := factory.TransactionTopic + ts.shardCoordinator.CommunicationIdentifier(senderShardId)
+
+	packets, err := ts.dataPacker.PackDataInChunks(transactions, common.MaxBulkTransactionSize)
+	if err != nil {
+		return err
+	}
+
+	for _, buff := range packets {
+		go func(bufferToSend []byte) {
+			log.Trace("txsSender.sendBulkTransactionsFromShard",
+				"topic", identifier,
+				"size", len(bufferToSend),
+			)
+			err = ts.networkMessenger.BroadcastOnChannelBlocking(
+				SendTransactionsPipe,
+				identifier,
+				bufferToSend,
+			)
+			log.LogIfError(err)
+		}(buff)
+	}
+
+	return nil
+}
+
+// printTxSentCounter prints the peak transaction counter from a time frame of about 'numSecondsBetweenPrints' seconds
+// if this peak value is 0 (no transaction was sent through the REST API interface), the print will not be done
+// the peak counter resets after each print. There is also a total number of transactions sent to p2p
+// TODO make this function testable. Refactor if necessary.
```

Comment:
will try to come-up with something better than this

Reply:
Yes, would be worth a refactor in the future.